parlai/core/dict.py

```diff
@@ -579,82 +586,69 @@ class _BPEHelper(object):
     in a second pass, calling tokenize() again to get processed output.
     """
 
-    def __init__(self, codecs_filename, num_symbols=30000, minfreq=2):
+    def __init__(self, codecs_filename):
         """
         Initialize the BPE module.
         If `codecs_filename` already exists, loads the pretrained codecs.
         If it does not, codecs will be saved there after a call to `finalize()`.
 
         :param codecs_filename: place to save/load codecs.
-        :param num_symbols: Number of BPE symbols. Recommend 30000-40000.
-        :param minfreq: Minimum frequency of a token before forced BPE decomposition.
         """
         if not BPE_INSTALLED:
             raise RuntimeError(
                 "Please run \"pip install 'git+https://github.com/rsennrich"
                 "/subword-nmt.git#egg=subword-nmt'\""
             )
 
+        self.splitter = re.compile(r'\w+|[^\w\s]', re.UNICODE)
+
         self.codecs = codecs_filename
         if os.path.exists(self.codecs):
-            self.built = True
             self._load_from_codecs()
-        else:
-            self.num_symbols = num_symbols
-            self.minfreq = minfreq
-            self.training_data = []
-            self.built = False
 
     def _load_from_codecs(self):
         with open(self.codecs, 'r') as codecs_file:
             self.bpe = apply_bpe.BPE(codecs_file)
 
-    def _add_to_train(self, tokens):
-        """
-        Queues up all the data to learn the BPE tokenizer.
-
-        :param tokens: list[str]. Initial tokenization approximation.
-        """
-        if self.built:
-            raise RuntimeError("BPE dictionary has been finalized.")
-        self.training_data.append(" ".join(tokens) + "\n")
-        return []
-
     def tokenize(self, text):
         """
-        Tokenizes the text if codecs are already finalized.
-        Otherwise, stores data for learning codecs.
+        Tokenizes the text with bpe if codecs are already finalized.
+        Otherwise, returns the regularly split tokens that will train the bpe.
 
         :param text: str. Raw text to tokenize.
-        :return: a list of tokens. List will be empty if not tokenized.
+        :return: a list of tokens. Will use BPE once finalized.
         """
-        tokens = DictionaryAgent.re_tokenize(text)
-        if self.built:
-            return self._apply(tokens)
-        else:
-            return self._add_to_train(tokens)
+        text = text.replace('\n', ' __newln__ ')
+        tokens = self.splitter.findall(text)
 
-    def _apply(self, tokens):
-        return self.bpe.segment_tokens(tokens)
+        if hasattr(self, 'bpe'):
+            return self.bpe.segment_tokens(tokens)
+        else:
+            return tokens
 
-    def finalize(self):
-        """Build the codecs"""
-        if self.built:
```

Comment:
I'm worried about this, b/c without testing this code, it looks to me like loading a model from disk will still cause the BPE encoder to retrain. This is a very expensive step, and so if the dict and codecs files are already built, we should use them.